instrumentation/RATIONALE.md

```diff
@@ -2,6 +2,100 @@
+as a siblings, as opposed to children of the previous callback (in this case
+a client).
+
+This may sound unintuitive to those thinking in terms of callback nesting depth,
+but having a consistent structure allows traces to appear similar regardless of
+imperative vs async invocation. It also is more easy to reason with, but we'll
+touch on that later.
+
+Let's consider the above async pseudo code with the logical equivalent in
+synchronous code. In each case, there are 3 client calls made in sequence. In
+each case, there's a potential data dependency, but it isn't actually used!
+```java
+// synchronous
+ScopedSpan parent = tracer.startScopedSpan("parent");
+try {
+  client.call("1");
+  client.call("2");
+  client.call("3");
+} finally {
+  parent.finish();
+}
+
+// reactive
+call = client.call("1")
+             .flatMap((r) -> client.call("2"))
+             .flatMap((r) -> client.call("3"));
+
+ScopedSpan parent = tracer.startScopedSpan("parent");
+try {
+  call.subscribe(subscriber);
+} finally {
+  parent.finish();
+}
+```
+
+We mention that data is ignored to highlight one deduction one can make, which
+is that the hierarchy should represent data dependency, as opposed to logical
+or time wise. While this is interesting, it is difficult to execute in
+practice. Instrumentation are usually at a lower level than the application
+code that they run. Hierarchy is already chosen before it would know if data
+would be read or not. In most cases, it would be unknowable if data were read
+or used at that level of abstraction. In other words, such a relationship is
+more fitting for span tags at a higher level, and decoupled from span
+hierarchy.
+
+Even throwing out the data dependency argument, some may think why not model
+callback depth anyway? We should model spans how the code looks, right?
+
+Only three calls may not seem that bad. Perhaps it is easy to reason with
+what's going on. However, what if there were 100 or 1000? It would be very
+difficult to reason with the actual parent which may be 999 levels up the tree.
+Some backend code perform operations like counting children, in order to
+determine fan out counts. This code would become useless as there would only
+ever be one child! Put visually, imagine clicking '-' 999 times to find the
+real parent in a typical trace UI!
+
+We acknowledge that using the invocation context as the parent of follow-up
+requests (response callback) is imperfect. It means any data dependency between
+one response and the next request is not represented in the hierarchy. It also
+means callback depth with not manifest in the trace hierarchy. That said,
+follow-up requests still share not just the same trace, but also the same local
+root, and also direct parent. As the clocks are the same (and in fact locked
+against skew), the happens after relationship manifests in span timing. At any
+rate, if a data dependency is important, you can consider mapping it as a tag.
```

Comment:
How about also talking about local spans (span without kind)? I guess some people want to model grouped client spans and intuitively think the group should also be a client span, though in our model, where we want to describe inter-service links, this should be a local span.

Reply:
sounds good will add

Reply:
I think it is a different topic, so I'll add a different section.

Is there a coherent link between this and that or just something that caught your mind. I'm not sure where to join the points as this is about the response callback and not the span kind vs no span kind topic..

I know we have another #1172 about that..

Reply:
moved to https://github.com/openzipkin/brave/pull/1172